### PR TITLE
simplified the round robin in RoutingTableView

### PR DESCRIPTION
**Pull request by pompon0**
also hidden some implementation details, which were exposed to PeerManagerActor.

The previous implementation of next hop selection was slightly favoring sending messages to newly connected peers for unknown reasons. The current one aims at uniform utilization of all connected peers and is way simpler. The expected impact on the actual performance is negligible.

### CODE REVIEW

File: chain/network/src/routing/routing_table_view.rs [hunk 1]
```diff
@@ -12,24 +11,28 @@ use std::sync::Arc;
...
-    pub(crate) peer_forwarding: Arc<HashMap<PeerId, Vec<PeerId>>>,
+    my_peer_id: PeerId,
     /// Store last update for known edges. This is limited to list of adjacent edges to `my_peer_id`.
-    pub(crate) local_edges_info: HashMap<PeerId, Edge>,
+    local_edges_info: HashMap<PeerId, Edge>,
+
+    /// Maps an account_id to a peer owning it.
+    account_peers: LruCache<AccountId, AnnounceAccount>,
+    /// Maps a peer_id to a set of connected peers that on the shortest path to peer_id.
+    routing_table: RoutingTable,
```
Comment (matklad) [inline chain/network/src/routing/routing_table_view.rs:26]: Looking at this, perhaps

```rust
next_hop: NextHopTable
```

would be a better name? "RoutingTable" to me sounds like it should contain full routes, and the types makes it look like it does, but in reality only next hop is stored I think?

File: chain/network/src/routing/routing_table_view.rs [hunk 2]
```diff
@@ -60,30 +63,27 @@ impl RoutingTableView {
...
+    /// Select a connected peer on some shortest path to `peer_id`.
     fn find_route_from_peer_id(&mut self, peer_id: &PeerId) -> Result<PeerId, FindRouteError> {
-        let routes = self.peer_forwarding.get(peer_id).ok_or(FindRouteError::PeerNotFound)?;
-        // Neighbor with minimum and maximum nonce respectively.
-        let ((min_v, next_hop), (max_v, _)) = (routes.iter())
-            .map(|peer_id| (self.route_nonce.get(peer_id).cloned().unwrap_or_default(), peer_id))
-            .minmax()
-            .into_option()
+        let peers = self.routing_table.get(peer_id).ok_or(FindRouteError::PeerNotFound)?;
+        // Strategy similar to Round Robin: select a viable peer which was least recently used.
```
[the remark this answers is not in the document]
Comment (matklad) [reply chain/network/src/routing/routing_table_view.rs:69]: Actually, let's just cut this out and specify, in the doc string, what exactly we are doing. Now that the algorithm is much simpler, this is possible: 

```
/// Select a connected peer on some shortest path to `peer_id`.
///
/// If there are several such peers, pick the least recently used one.
```

File: chain/network/src/routing/routing_table_view.rs [hunk 3]
```diff
@@ -12,24 +11,28 @@ use std::sync::Arc;
 use tracing::warn;
 
 const ANNOUNCE_ACCOUNT_CACHE_SIZE: usize = 10_000;
-const ROUND_ROBIN_MAX_NONCE_DIFFERENCE_ALLOWED: usize = 10;
-const ROUND_ROBIN_NONCE_CACHE_SIZE: usize = 10_000;
+const LAST_ROUTED_CACHE_SIZE: usize = 10_000;
+
+type RoutingTable = Arc<HashMap<PeerId, Vec<PeerId>>>;
```
Comment (matklad) [inline chain/network/src/routing/routing_table_view.rs:16]: We have another `pub type RoutingTable = HashMap<PeerId, Vec<PeerId>>;` in `graph_with_cache`. Could use that here given the `pub`? It also seems fit to let the `Arc` outside

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 4]
```diff
@@ -356,8 +356,10 @@ impl PeerManagerActor {
                     routing_table,
                     peers_to_ban,
                 }) => {
-                    act.routing_table_view.remove_local_edges(local_edges_to_remove.iter());
-                    act.routing_table_view.peer_forwarding = routing_table.clone();
+                    for peer_id in &local_edges_to_remove {
+                        act.routing_table_view.remove_local_edge(peer_id);
+                    }
+                    act.routing_table_view.set_routing_table(routing_table.clone());
```
[the remark this answers is not in the document]
Comment (matklad) [reply chain/network/src/peer_manager/peer_manager_actor.rs:362]: Or maybe not: taking a second look here, `local_edge` stuff seems orthogonal to routing_table

**Review by matklad — APPROVED** [review]
LGTM!

One thing I am not sure if the high-level change of logic is good. It does seem good, up to the point of me wondering why didn't we do this in the first place, which makes me wonder if I am missing something. 

Was it perhaps just historical development? Did we initially tracked only local nonces, then noticed that it allows a newly introduced node to eclipse commpunciation and added the max-nonce-difference stuff?

**Comment by matklad** [comment]
Another observation: it seems no tests were updated. Do we have any check that we actually round-robin here?

File: chain/network/src/routing/routing_table_view.rs (hunk 1 above)
Comment (matklad) [reply chain/network/src/routing/routing_table_view.rs:26]: (even if it is, def out of scope for this PR though!)

File: chain/network/src/routing/routing_table_view.rs [hunk 5]
```diff
@@ -12,24 +11,28 @@ use std::sync::Arc;
...
     store: store::Store,
-    /// Number of times each active connection was used to route a message.
-    /// If there are several options use route with minimum nonce.
-    /// New routes are added with minimum nonce.
-    route_nonce: LruCache<PeerId, usize>,
+
+    /// Counter of number of calls to find_route_by_peer_id.
+    find_route_calls: usize,
+    /// Last time the given peer was selected by find_route_by_peer_id.
+    last_routed: LruCache<PeerId, usize>,
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs (hunk 3 above)
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs [hunk 6]
```diff
@@ -198,10 +192,23 @@ impl RoutingTableView {
...
+    pub(crate) fn remove_local_edge(&mut self, peer_id: &PeerId) {
+        self.local_edges_info.remove(peer_id);
+    }
 }
 
 #[derive(Debug)]
 pub struct RoutingTableInfo {
     pub account_peers: HashMap<AccountId, PeerId>,
-    pub peer_forwarding: Arc<HashMap<PeerId, Vec<PeerId>>>,
+    pub routing_table: Arc<HashMap<PeerId, Vec<PeerId>>>,
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs [hunk 7]
```diff
@@ -60,30 +63,27 @@ impl RoutingTableView {
...
+        let peers = self.routing_table.get(peer_id).ok_or(FindRouteError::PeerNotFound)?;
+        // Strategy similar to Round Robin: select a viable peer which was least recently used.
+        let next_hop = peers
+            .iter()
+            .min_by_key(|p| self.last_routed.get(p.clone()).cloned().unwrap_or(0))
             .ok_or(FindRouteError::Disconnected)?;
-        // Strategy similar to Round Robin. Select node with least nonce and send it. Increase its
-        // nonce by one. Additionally if the difference between the highest nonce and the lowest
-        // nonce is greater than some threshold increase the lowest nonce to be at least
-        // max nonce - threshold.
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs:76]: done

File: chain/network/src/routing/routing_table_view.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs [hunk 8]
```diff
@@ -60,30 +63,27 @@ impl RoutingTableView {
...
-    /// from `source` to `peer_id`.
+    /// Select a connected peer on some shortest path to `peer_id`.
     fn find_route_from_peer_id(&mut self, peer_id: &PeerId) -> Result<PeerId, FindRouteError> {
-        let routes = self.peer_forwarding.get(peer_id).ok_or(FindRouteError::PeerNotFound)?;
-        // Neighbor with minimum and maximum nonce respectively.
-        let ((min_v, next_hop), (max_v, _)) = (routes.iter())
-            .map(|peer_id| (self.route_nonce.get(peer_id).cloned().unwrap_or_default(), peer_id))
-            .minmax()
-            .into_option()
+        let peers = self.routing_table.get(peer_id).ok_or(FindRouteError::PeerNotFound)?;
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs [hunk 9]
```diff
@@ -60,30 +63,27 @@ impl RoutingTableView {
...
+        // Strategy similar to Round Robin: select a viable peer which was least recently used.
+        let next_hop = peers
+            .iter()
+            .min_by_key(|p| self.last_routed.get(p.clone()).cloned().unwrap_or(0))
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

File: chain/network/src/routing/routing_table_view.rs [hunk 10]
```diff
@@ -12,24 +11,28 @@ use std::sync::Arc;
...
-    /// Active PeerId that are part of the shortest path to each PeerId.
-    pub(crate) peer_forwarding: Arc<HashMap<PeerId, Vec<PeerId>>>,
+    my_peer_id: PeerId,
     /// Store last update for known edges. This is limited to list of adjacent edges to `my_peer_id`.
-    pub(crate) local_edges_info: HashMap<PeerId, Edge>,
+    local_edges_info: HashMap<PeerId, Edge>,
+
+    /// Maps an account_id to a peer owning it.
+    account_peers: LruCache<AccountId, AnnounceAccount>,
+    /// Maps a peer_id to a set of connected peers that on the shortest path to peer_id.
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/routing/routing_table_view.rs]: done

**Pull request merged**